### ISSUE TITLE
A way to NOOP http requests in non-production env

### ISSUE THREAD

**Issue by rauanmayemir** [bugsnag/bugsnag-laravel#362]
### Expected behavior
I would like to be able to simply register `BugsnagServiceProvider` in all environments and use things like `Bugsnag::notifyError` and `Bugsnag::notifyException` in my app logic. So if it's a test or local env, such calls should simply be ignored.

### Observed behavior

My app in test env fail due to not being able to connect to bugsnag server. Makes sense, I don't set api key in test env.

**Comment by GrahamCampbell** [comment]
Yes. You can use the `notify_release_stages` config option. :)

**Comment by GrahamCampbell** [comment]
https://docs.bugsnag.com/platforms/php/laravel/configuration-options/#release-stage

**Comment by rauanmayemir** [comment]
I tried it before and just tried again, that doesn't seem to work. I even tried setting it directly to something like `'notify_release_stages' => ['production'],` in `config/bugsnag.php`, but the result is always same: `phpunit` call fails with `"Missing 'Bugsnag-Api-Key' header"`.

Is there a minimal basic setup to force bugsnag connect to the server only in production env?

**Comment by GrahamCampbell** [comment]
Could you show the stack trace for that error please?

**Comment by rauanmayemir** [comment]
```
[2019-08-26 21:48:03] testing.ERROR: Client error: `POST https://sessions.bugsnag.com` resulted in a `400 Bad Request` response:
{"status":"error","errors":["Missing 'Bugsnag-Api-Key' header"]}
 {"userId":1,"exception":"[object] (GuzzleHttp\\Exception\\ClientException(code: 400): Client error: `POST https://sessions.bugsnag.com` resulted in a `400 Bad Request` response:
{\"status\":\"error\",\"errors\":[\"Missing 'Bugsnag-Api-Key' header\"]}
 at app/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php:113)
[stacktrace]
#0 app/vendor/guzzlehttp/guzzle/src/Middleware.php(66): GuzzleHttp\\Exception\\RequestException::create(Object(GuzzleHttp\\Psr7\\Request), Object(GuzzleHttp\\Psr7\\Response))
#1 app/vendor/guzzlehttp/promises/src/Promise.php(203): GuzzleHttp\\Middleware::GuzzleHttp\\{closure}(Object(GuzzleHttp\\Psr7\\Response))
#2 app/vendor/guzzlehttp/promises/src/Promise.php(156): GuzzleHttp\\Promise\\Promise::callHandler(1, Object(GuzzleHttp\\Psr7\\Response), Array)
#3 app/vendor/guzzlehttp/promises/src/TaskQueue.php(47): GuzzleHttp\\Promise\\Promise::GuzzleHttp\\Promise\\{closure}()
#4 app/vendor/guzzlehttp/promises/src/Promise.php(246): GuzzleHttp\\Promise\\TaskQueue->run(true)
#5 app/vendor/guzzlehttp/promises/src/Promise.php(223): GuzzleHttp\\Promise\\Promise->invokeWaitFn()
#6 app/vendor/guzzlehttp/promises/src/Promise.php(267): GuzzleHttp\\Promise\\Promise->waitIfPending()
#7 app/vendor/guzzlehttp/promises/src/Promise.php(225): GuzzleHttp\\Promise\\Promise->invokeWaitList()
#8 app/vendor/guzzlehttp/promises/src/Promise.php(62): GuzzleHttp\\Promise\\Promise->waitIfPending()
#9 app/vendor/guzzlehttp/guzzle/src/Client.php(131): GuzzleHttp\\Promise\\Promise->wait()
#10 app/vendor/guzzlehttp/guzzle/src/Client.php(89): GuzzleHttp\\Client->request('post', '', Array)
#11 app/vendor/bugsnag/bugsnag/src/SessionTracker.php(300): GuzzleHttp\\Client->__call('post', Array)
#12 app/vendor/bugsnag/bugsnag/src/SessionTracker.php(228): Bugsnag\\SessionTracker->deliverSessions()
#13 app/vendor/bugsnag/bugsnag/src/SessionTracker.php(125): Bugsnag\\SessionTracker->incrementSessions('2019-08-26T21:4...')
#14 app/vendor/bugsnag/bugsnag-laravel/src/BugsnagServiceProvider.php(87): Bugsnag\\SessionTracker->startSession()
#15 app/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(347): Bugsnag\\BugsnagLaravel\\BugsnagServiceProvider->Bugsnag\\BugsnagLaravel\\{closure}(Object(Illuminate\\Routing\\Events\\RouteMatched))
```

**Comment by GrahamCampbell** [comment]
Thanks. You you add `var_dump(config('bugsnag'))` (and remove your api key from the output), to your tests please.

Also, could you show more lines to the trace please?

**Comment by rauanmayemir** [comment]
Here's the config:

```
array(21) {
  ["api_key"]=>
  string(0) ""
  ["app_type"]=>
  NULL
  ["app_version"]=>
  string(8) "51cabb74"
  ["batch_sending"]=>
  NULL
  ["endpoint"]=>
  NULL
  ["filters"]=>
  array(1) {
    [0]=>
    string(8) "password"
  }
  ["hostname"]=>
  NULL
  ["proxy"]=>
  array(0) {
  }
  ["project_root"]=>
  NULL
  ["strip_path"]=>
  NULL
  ["query"]=>
  bool(true)
  ["bindings"]=>
  bool(true)
  ["release_stage"]=>
  NULL
  ["notify_release_stages"]=>
  array(2) {
    [0]=>
    string(7) "staging"
    [1]=>
    string(10) "production"
  }
  ["send_code"]=>
  bool(true)
  ["callbacks"]=>
  bool(true)
  ["user"]=>
  bool(true)
  ["logger_notify_level"]=>
  NULL
  ["auto_capture_sessions"]=>
  bool(true)
  ["session_endpoint"]=>
  NULL
  ["build_endpoint"]=>
  NULL
}
```

And here's the full stacktrace:

```
[2019-08-26 22:38:50] testing.ERROR: Client error: `POST https://sessions.bugsnag.com` resulted in a `400 Bad Request` response:
{"status":"error","errors":["Missing 'Bugsnag-Api-Key' header"]}
 {"userId":1,"exception":"[object] (GuzzleHttp\\Exception\\ClientException(code: 400): Client error: `POST https://sessions.bugsnag.com` resulted in a `400 Bad Request` response:
{\"status\":\"error\",\"errors\":[\"Missing 'Bugsnag-Api-Key' header\"]}
 at app/vendor/guzzlehttp/guzzle/src/Exception/RequestException.php:113)
[stacktrace]
#0 app/vendor/guzzlehttp/guzzle/src/Middleware.php(66): GuzzleHttp\\Exception\\RequestException::create(Object(GuzzleHttp\\Psr7\\Request), Object(GuzzleHttp\\Psr7\\Response))
#1 app/vendor/guzzlehttp/promises/src/Promise.php(203): GuzzleHttp\\Middleware::GuzzleHttp\\{closure}(Object(GuzzleHttp\\Psr7\\Response))
#2 app/vendor/guzzlehttp/promises/src/Promise.php(156): GuzzleHttp\\Promise\\Promise::callHandler(1, Object(GuzzleHttp\\Psr7\\Response), Array)
#3 app/vendor/guzzlehttp/promises/src/TaskQueue.php(47): GuzzleHttp\\Promise\\Promise::GuzzleHttp\\Promise\\{closure}()
#4 app/vendor/guzzlehttp/promises/src/Promise.php(246): GuzzleHttp\\Promise\\TaskQueue->run(true)
#5 app/vendor/guzzlehttp/promises/src/Promise.php(223): GuzzleHttp\\Promise\\Promise->invokeWaitFn()
#6 app/vendor/guzzlehttp/promises/src/Promise.php(267): GuzzleHttp\\Promise\\Promise->waitIfPending()
#7 app/vendor/guzzlehttp/promises/src/Promise.php(225): GuzzleHttp\\Promise\\Promise->invokeWaitList()
#8 app/vendor/guzzlehttp/promises/src/Promise.php(62): GuzzleHttp\\Promise\\Promise->waitIfPending()
#9 app/vendor/guzzlehttp/guzzle/src/Client.php(131): GuzzleHttp\\Promise\\Promise->wait()
#10 app/vendor/guzzlehttp/guzzle/src/Client.php(89): GuzzleHttp\\Client->request('post', '', Array)
#11 app/vendor/bugsnag/bugsnag/src/SessionTracker.php(300): GuzzleHttp\\Client->__call('post', Array)
#12 app/vendor/bugsnag/bugsnag/src/SessionTracker.php(228): Bugsnag\\SessionTracker->deliverSessions()
#13 app/vendor/bugsnag/bugsnag/src/SessionTracker.php(125): Bugsnag\\SessionTracker->incrementSessions('2019-08-26T22:3...')
#14 app/vendor/bugsnag/bugsnag-laravel/src/BugsnagServiceProvider.php(87): Bugsnag\\SessionTracker->startSession()
#15 app/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(347): Bugsnag\\BugsnagLaravel\\BugsnagServiceProvider->Bugsnag\\BugsnagLaravel\\{closure}(Object(Illuminate\\Routing\\Events\\RouteMatched))
#16 app/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(196): Illuminate\\Events\\Dispatcher->Illuminate\\Events\\{closure}('Illuminate\\\\Rout...', Array)
#17 app/vendor/fntneves/laravel-transactional-events/src/Neves/Events/TransactionalDispatcher.php(114): Illuminate\\Events\\Dispatcher->dispatch('Illuminate\\\\Rout...', Array, false)
#18 app/vendor/laravel/framework/src/Illuminate/Routing/Router.php(654): Neves\\Events\\TransactionalDispatcher->dispatch(Object(Illuminate\\Routing\\Events\\RouteMatched))
#19 app/vendor/laravel/framework/src/Illuminate/Routing/Router.php(623): Illuminate\\Routing\\Router->runRoute(Object(Dingo\\Api\\Http\\Request), Object(Illuminate\\Routing\\Route))
#20 app/vendor/laravel/framework/src/Illuminate/Routing/Router.php(612): Illuminate\\Routing\\Router->dispatchToRoute(Object(Dingo\\Api\\Http\\Request))
#21 app/vendor/dingo/api/src/Routing/Adapter/Laravel.php(88): Illuminate\\Routing\\Router->dispatch(Object(Dingo\\Api\\Http\\Request))
#22 app/vendor/dingo/api/src/Routing/Router.php(514): Dingo\\Api\\Routing\\Adapter\\Laravel->dispatch(Object(Dingo\\Api\\Http\\Request), 'v1')
#23 app/vendor/dingo/api/src/Http/Middleware/Request.php(126): Dingo\\Api\\Routing\\Router->dispatch(Object(Dingo\\Api\\Http\\Request))
#24 app/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Dingo\\Api\\Http\\Middleware\\Request->Dingo\\Api\\Http\\Middleware\\{closure}(Object(Dingo\\Api\\Http\\Request))
#25 app/vendor/fideloper/proxy/src/TrustProxies.php(57): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Dingo\\Api\\Http\\Request))
#26 app/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Fideloper\\Proxy\\TrustProxies->handle(Object(Dingo\\Api\\Http\\Request), Object(Closure))
#27 app/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Dingo\\Api\\Http\\Request))
#28 app/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle(Object(Dingo\\Api\\Http\\Request), Object(Closure))
#29 app/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Dingo\\Api\\Http\\Request))
#30 app/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle(Object(Dingo\\Api\\Http\\Request), Object(Closure))
#31 app/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Dingo\\Api\\Http\\Request))
#32 app/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle(Object(Dingo\\Api\\Http\\Request), Object(Closure))
#33 app/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php(62): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Dingo\\Api\\Http\\Request))
#34 app/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\\Foundation\\Http\\Middleware\\CheckForMaintenanceMode->handle(Object(Dingo\\Api\\Http\\Request), Object(Closure))
#35 app/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(104): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Dingo\\Api\\Http\\Request))
#36 app/vendor/dingo/api/src/Http/Middleware/Request.php(127): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
#37 app/vendor/dingo/api/src/Http/Middleware/Request.php(103): Dingo\\Api\\Http\\Middleware\\Request->sendRequestThroughRouter(Object(Dingo\\Api\\Http\\Request))
#38 app/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Dingo\\Api\\Http\\Middleware\\Request->handle(Object(Dingo\\Api\\Http\\Request), Object(Closure))
#39 app/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#40 app/vendor/barryvdh/laravel-cors/src/HandlePreflight.php(29): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))
#41 app/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Barryvdh\\Cors\\HandlePreflight->handle(Object(Illuminate\\Http\\Request), Object(Closure))
#42 app/vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}(Object(Illuminate\\Http\\Request))
#43 app/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(104): Illuminate\\Routing\\Pipeline->Illuminate\\Routing\\{closure}(Object(Illuminate\\Http\\Request))
#44 app/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(151): Illuminate\\Pipeline\\Pipeline->then(Object(Closure))
#45 app/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(116): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter(Object(Illuminate\\Http\\Request))
#46 app/vendor/laravel/framework/src/Illuminate/Foundation/Testing/Concerns/MakesHttpRequests.php(375): Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))
#47 app/vendor/laravel/framework/src/Illuminate/Foundation/Testing/Concerns/MakesHttpRequests.php(197): Illuminate\\Foundation\\Testing\\TestCase->call('POST', '/api/v1/account...', Array, Array, Array, Array)
#48 app/tests/MyTest.php(78): Illuminate\\Foundation\\Testing\\TestCase->post('/api/v1/account...', Array)
#49 phar:///usr/local/Cellar/phpunit/7.4.3/bin/phpunit/phpunit/Framework/TestCase.php(1150): Tests\\MyTest->testCreated()
#50 phar:///usr/local/Cellar/phpunit/7.4.3/bin/phpunit/phpunit/Framework/TestCase.php(844): PHPUnit\\Framework\\TestCase->runTest()
#51 phar:///usr/local/Cellar/phpunit/7.4.3/bin/phpunit/phpunit/Framework/TestResult.php(675): PHPUnit\\Framework\\TestCase->runBare()
#52 phar:///usr/local/Cellar/phpunit/7.4.3/bin/phpunit/phpunit/Framework/TestCase.php(798): PHPUnit\\Framework\\TestResult->run(Object(Tests\\MyTest))
#53 phar:///usr/local/Cellar/phpunit/7.4.3/bin/phpunit/phpunit/Framework/TestSuite.php(750): PHPUnit\\Framework\\TestCase->run(Object(PHPUnit\\Framework\\TestResult))
#54 phar:///usr/local/Cellar/phpunit/7.4.3/bin/phpunit/phpunit/TextUI/TestRunner.php(622): PHPUnit\\Framework\\TestSuite->run(Object(PHPUnit\\Framework\\TestResult))
#55 phar:///usr/local/Cellar/phpunit/7.4.3/bin/phpunit/phpunit/TextUI/Command.php(206): PHPUnit\\TextUI\\TestRunner->doRun(Object(PHPUnit\\Framework\\TestSuite), Array, true)
#56 phar:///usr/local/Cellar/phpunit/7.4.3/bin/phpunit/phpunit/TextUI/Command.php(162): PHPUnit\\TextUI\\Command->run(Array, true)
#57 /usr/local/Cellar/phpunit/7.4.3/bin/phpunit(616): PHPUnit\\TextUI\\Command::main()
#58 {main}
"}
```

**Comment by GrahamCampbell** [comment]
Ok, thanks for the information. It looks like sessions aren't respecting the release stage config.

**Comment by mattdyoung** [comment]
Fixed in bugsnag-php [v3.19.0](https://github.com/bugsnag/bugsnag-php/releases/tag/v3.19.0)
